_posts/2016-03-14-make-your-data-obey.markdown

```diff
@@ -0,0 +1,223 @@
+
+## Modifiers
+
+Obey wasn't just built for validation, but also to support coercion and modification. Modifiers allow for doing this in a clean, simple way. Let's revisit the model again, using the `modifier` method to create one:
+
+{% highlight javascript %}
+import obey from 'obey'
+import uuid from 'node-uuid'
+// Import the argon2 module
+import argon2 from 'argon2'
+
+obey.creator('uuidCreator', uuid.v4)
+
+// Add a modifier that returns md5 encrypted value
+const salt = new Buffer('somesalt')
+obey.modifier('encrypt', val => argon2.hash(val, salt))
+
+const user = obey.model({
+    id: { type: 'uuid', creator: 'uuidCreator' required: true },
+    email: { type: 'email', required: true },
+    // Password now specifies the modifier 'encrypt' should be used to
+    // encrypt the password
+    password: { type: 'string', modifier: 'encrypt', require: true }
+})
+{% endhighlight %}
+
+Now not only does our validation method create a UUID, it also encrypts the password for us.
+
+But what about that `email` property? Probably want to make sure it's unique...
+
+## Types
+
+Having that `email` property contain one that's already in the datasource would be bad news, right? Some databases do provide uniqueness integrity checking, but there are a lot that don't. Luckily, `types` are extensible with Obey, so we can add our own check:
+
+{% highlight javascript %}
+import obey from 'obey'
+import uuid from 'node-uuid'
+import argon2 from 'argon2'
+// Here's our db reference again...
+import db from './db'
+
+obey.creator('uuidCreator', uuid.v4)
+
+const salt = new Buffer('somesalt')
+obey.modifier('encrypt', val => argon2.hash(val, salt))
+
+// Add a type to ensure we have a valid, unique email
+obey.type('uniqueEmail', context => {
+    // Read from the 'db' datasource where email is the value
+    return db.read({ email: context.val }).then(res => {
+        // Fail if it already exists
+        if (res.length !== 0) context.fail('Email must be unique')
+        // Note: you'd also want to check that the email is valid!
+    })
+})
+
+const user = obey.model({
+    id: { type: 'uuid', creator: 'uuidCreator' required: true },
+    // Email is now set to type 'uniqueEmail'
+    email: { type: 'uniqueEmail', required: true },
+    password: { type: 'string', modifier: 'encrypt', require: true }
+})
+{% endhighlight %}
+
+Now when the `validate` method is run we do a read on the datasource and check that the email doesn't already exist. Obey doesn't need to be returned a value or `true`, if the `context.fail` method is not called it will simply move forward, adding the error to a collection which is included with any other errors.
+
+What's best is **we do all of this at the data validation level**. Yes, we're still hitting the datasource (really no way around that), but all of our handling of an error condition is in the same place as the rest of the validation error handling.
```

Comment:
```
but all of our error handling conditions are in the same place as the rest of the validation handling
```
might be easier to decipher than 

```
but all of our handling of an error condition is in the same place as the rest of the validation error handling
```

Note, I dropped _error_ from the last part of that sentence because it almost seemed redundant, or perhaps confusing as to why we would mention having conditions for error handling along side our actual error handling. But mentioning that we have the error handling along side the validation handling (the core of Obey) still seemed appropriate and seems to draw a bigger _wow factor_ to me.